pyop2/mpi.py

```diff
@@ -118,63 +135,272 @@ def delcomm_outer(comm, keyval, icomm):
-compilationcomm_keyval = MPI.Comm.Create_keyval()
+@collective
+def create_split_comm(comm):
+    if MPI.VERSION >= 3:
+        debug("Creating compilation communicator using MPI_Split_type")
+        split_comm = comm.Split_type(MPI.COMM_TYPE_SHARED)
+        debug("Finished creating compilation communicator using MPI_Split_type")
+    else:
+        debug("Creating compilation communicator using MPI_Split + filesystem")
+        import tempfile
+        if comm.rank == 0:
+            if not os.path.exists(configuration["cache_dir"]):
+                os.makedirs(configuration["cache_dir"], exist_ok=True)
+            tmpname = tempfile.mkdtemp(prefix="rank-determination-",
+                                       dir=configuration["cache_dir"])
+        else:
+            tmpname = None
+        tmpname = comm.bcast(tmpname, root=0)
+        if tmpname is None:
+            raise CompilationError("Cannot determine sharedness of filesystem")
+        # Touch file
+        debug("Made tmpdir %s" % tmpname)
+        with open(os.path.join(tmpname, str(comm.rank)), "wb"):
+            pass
+        comm.barrier()
+        import glob
+        ranks = sorted(int(os.path.basename(name))
+                       for name in glob.glob("%s/[0-9]*" % tmpname))
+        debug("Creating compilation communicator using filesystem colors")
+        split_comm = comm.Split(color=min(ranks), key=comm.rank)
+        debug("Finished creating compilation communicator using filesystem colors")
+    # Name
+    split_comm.Set_name(f"{comm.name or id(comm)}_COMPILATION")
+    # Refcount
+    split_comm.Set_attr(refcount_keyval, [0])
+    incref(split_comm)
+    return split_comm
 
 
 def get_compilation_comm(comm):
     return comm.Get_attr(compilationcomm_keyval)
 
 
 def set_compilation_comm(comm, inner):
-    comm.Set_attr(compilationcomm_keyval, inner)
+    """Set the compilation communicator.
+
+    :arg comm: A PyOP2 Communicator
+    :arg inner: The compilation communicator
+    """
+    # Ensure `comm` is a PyOP2 comm
+    if not is_pyop2_comm(comm):
+        raise ValueError("Compilation communicator must be stashed on a PyOP2 comm")
+
+    # Check if the compilation communicator is already set
+    old_inner = comm.Get_attr(compilationcomm_keyval)
+    if old_inner is not None:
+        if is_pyop2_comm(old_inner):
+            raise ValueError("Compilation communicator is not a PyOP2 comm, something is very broken!")
+        else:
+            decref(old_inner)
+
+    if not is_pyop2_comm(inner):
+        raise ValueError(
+            "Communicator used for compilation communicator must be a PyOP2 communicator.\n"
+            "Use pyop2.mpi.dup_comm() to create a PyOP2 comm from an existing comm.")
+    else:
+        # Stash `inner` as an attribute on `comm`
+        comm.Set_attr(compilationcomm_keyval, inner)
 
 
-def free_comm(comm, remove=True):
+@collective
+def compilation_comm(comm):
```

Comment:
Why `set_compilation_comm` and `compilation_comm`?

Reply:
Bit of a left over from previous state of code. Basically `compilation_comm` works like `internal_comm` in that it creates a new communicator only if necessary. `set_compilation_comm` is just stashing one comm on another, but this needs reference counting now!